### PR TITLE
Fix/pr 460 bunysae - fix clicking an item on mobile

### LINKED ISSUE #181: click to change - unable to select a photo for a low width of the screen
[Issuehunt badges](https://issuehunt.io/r/brainhubeu/react-carousel/issues/181)

**To Reproduce**
1. Open https://brainhubeu.github.io/react-carousel/docs/examples/clickToChange with Safari on iOS (iPhone) so there are three photos and the code is following:
```js
<Carousel
  clickToChange
  slidesPerPage={2}
  centered
>
 <img src={imageOne} />
 <img src={imageTwo} />
 <img src={imageThree} />
</Carousel>
```
2. Set the screen to be vertical (default for iPhone) so the screen width is very low.
3. Try to click on the most left photo.
4. It's impossible to select the most left photo.
5. Click the most right photo.
6. The most right photo occupies all the carousel.
7. So the only way to show the center photo is to drag.
8. After dragging, the center photo is shown but not fully.

**Expected behavior**
4. It should be possible to select the most left photo.
6. The most right photo should occupy most of the carousel but there should be also a space for the center photo, enabling clicking it.
7. It should be possible to show the center photo not only with dragging but also with clicking it.
8. The center photo should be shown fully, with some space for the most left and the most right photo, enabling clicking them.

No photo should be cut or shrunk horizontally.

Instead, the height of the entire carousel should be decreased when needed in order to fit always the selected image and parts of its neighbors, enabling to click them.

**Screenshots**
Video: https://drive.google.com/file/d/1gkDYx8qPymJh8PKLjNQ3UMp39n6vd_B4/view

**Smartphone (please complete the following information):**
 - Device: iPhone 5s
 - OS: iOS 12.4.4
 - Browser: Safari (AFAIK Safari version is connected with the iOS version)
 - `react-carousel`: 1.10.34

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $17.00)

- [piotr-s-brainhub piotr-s-brainhub](https://issuehunt.io/u/piotr-s-brainhub) ($17.00)

### Submitted pull Requests
- [#464 Fixes #181](https://issuehunt.io/r/brainhubeu/react-carousel/pull/464)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/brainhubeu/react-carousel/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by piotr-s-brainhub**
Deployed to https://beghp.github.io/gh-pages-rc-2<br>
It fixes #181
<br>- [x] an issue linked to the PR<br><br><br><br><br><br><br><br><br><br><br><br>

### CODE REVIEW

**Comment by piotr-s-brainhub** [comment]
@bunysae

It doesn't work correctly.
In #181 I wrote `No photo should be cut or shrunk horizontally.` but they are shrunk. They should be resized instead.
Video: https://drive.google.com/file/d/1lSYyW5KRY6u1yxQkD38yu0Gs0N2LpBM-/view?usp=sharing

How difficult would be to resize images in order to show them fully?
If it's not that simple, I can consider creating a separate issue.

**Comment by piotr-s-brainhub** [comment]
Moreover, this doesn't look good:
Screenshot 2020-05-01 at 07 22 53
Screenshot 2020-05-05 at 18 38 23
Screenshot 2020-05-05 at 18 38 37
http://g.recordit.co/nFVfQCiAoA.gif

**Comment by bunysae** [comment]
When the CSS properties `width`, `min-width`, `max-width` and `display: flex` are directly applied to the img tag, no image shrunks or is cut on iOS. 

> Moreover, this doesn't look good:

Should the images be centered?

**Comment by piotr-s-brainhub** [comment]
@bunysae 

so if you're able to fix that, please open a PR from your fork to the `fix/pr-460-bunysae-master` branch

**Comment by piotr-s-brainhub** [comment]
Now it works nice.
Video: https://drive.google.com/file/d/1T1rKfpIS-7mVjD0DXmb-zYq3ixEtewAq/view

However, there are some other issues:
Default arrows and custom arrows:
It is:
1) I open a given view
2) I click the right arrow
3) It moves immediately
4) I click the right arrow again
5) It moves with an animation
6) It moves an animation each time, for both right and left arrow and no matter which slide is currently selected
7) I change view
8) I go back to the given (default arrows and custom arrows) view
9) I click the right arrow
10) It moves immediately

It should be:
3) It should move with an animation
10) It should move with an animation
Issue video: https://recordit.co/amKno4MUhc

**Comment by bunysae** [comment]
I fixed it.

**Comment by piotr-s-brainhub** [comment]
@bunysae 

Thanks, now on the desktop everything is working but there's another mobile issue:

video: https://drive.google.com/file/d/1Px_uDQud480IfVOXQPklgmf9l8OglHFW/view

As you can see, it jumps in autoplay, slides per scroll and centered examples.

**Comment by bunysae** [comment]
I found a fix for iOS 13.4 using the `ResizeObserver`-API. For iOS 12.4 i got either the problem of jumping or a missing animation, when the infinite-option is used.

**Comment by piotr-s-brainhub** [comment]
@bunysae 

Now it works correctly on desktop and better on mobile.
However, in some examples, I see up-down blinking when changing slides (mainly in the right direction).

Video: https://drive.google.com/file/d/1vjnycLuBlowSK3GGWpA1a-9rrr_IVHw7/view

**Comment by piotr-s-brainhub** [comment]
@bunysae 

Are you going to fix this?

**Comment by bunysae** [comment]
I will try it. The main problem was when clicked right, new components
are mounted with their initial size.
In iOS 12 a polyfill must be used to observe the width. This polyfill
does react with a little delay. The native implementation of
ResizeObserver in iOS 13 doesn't have such a delay and therefore
don't have this problem.

**Comment by piotr-s-brainhub** [comment]
@bunysae 

Now I see, up-down blinking is not so frequent as it was previously but sometimes, it still happens.

Moreover, in the `Multiple items per page` example, mobile (iOS), when switching slides right, I noticed sometimes, it moved immediately with no animation.

**Comment by bunysae** [comment]
I found out, that iOS safari loads the images everytime a slide changes (infinite and autoplay example) from the server. This causes the blinking.

**Comment by piotr-s-brainhub** [comment]
@bunysae 

Do you think this https://github.com/brainhubeu/react-carousel/issues/389 will fix the blinking issue?

**Comment by bunysae** [comment]
@piotr-s-brainhub
Yes. My idea to fix the blinking issue and issue #389 would be
rethinking the additional clones right. At the moment 4 clones are
attached, when the slide is changed right. With the next render cycle
3 clones are attached. Theoretically i would be enough, to attach 1
clone to right. So we have more than enough clones on the right side.
Why we don't attach everytime 3 clones to the right?
I tried it and the blinking issue was gone and the images weren't
loaded multiple times.
What do you think about this idea?

**Comment by piotr-s-brainhub** [comment]
@bunysae 

IMO the simplest would be initially to create all the slide components, loading all the images and then use the same components instead of creating new ones

**Comment by bunysae** [comment]
Issue #389 is fixed with it, too. In my tests the images weren't loaded everytime, when the slide moved to the right.

**Comment by piotr-s-brainhub** [comment]
@bunysae 

Now, I cannot see the blinking (jumping up-down) issue anymore, thanks.

Regarding jumping left-right, I created this issue https://github.com/brainhubeu/react-carousel/issues/503 because it happens also on master.

Unfortunately, I see an issue on desktop with this PR (which doesn't occur on master): in the infinite example, when I move fast many slides right, the carousel disappears for a moment or sometimes it switches immediately to another slide. Video: https://recordit.co/3zxc69CMJt

**Comment by piotr-s-brainhub** [comment]
@bunysae 

Are you going to fix the issue about disappearing images when switching fast (https://recordit.co/3zxc69CMJt)?

**Comment by piotr-s-brainhub** [comment]
Thanks @bunysae 

Now it works fine.

I still see:
- jumping sometimes in `slides per scroll` on mobile but it already happens on `master` (https://github.com/brainhubeu/react-carousel/issues/503)
- blinking in `slides per scroll` on mobile but it looks better than the current version with images significantly flattened horizontally and maybe #389 will fix it

File: src/components/Carousel.js [hunk 1]
```diff
@@ -486,6 +490,7 @@ class Carousel extends Component {
     return dragOffset - currentValue * elementWidthWithOffset + additionalOffset - additionalClonesOffset;
   };
 
+  getIdCarouselItem = index => index % this.getChildren().length + 1;
```
Comment (RobertHebel) [inline src/components/Carousel.js:493]: Why do we use `this.getChildren().length + 1`?

File: src/components/CarouselItem.js [hunk 2]
```diff
@@ -12,11 +13,50 @@ class CarouselItem extends PureComponent {
...
+  /* ========== Resize, if necessary. Workaround for iOS safari ========== */
+  componentDidMount() {
+    this.observeWidth();
+  }
+
+  componentDidUpdate(prevProps) {
+    if (prevProps.width !== this.props.width) {
+      this.resizeChildren();
+    }
+  }
```
Comment (RobertHebel) [inline src/components/CarouselItem.js:36]: Does this method do the same as `observeWidth`? Do we need it?

Comment (bunysae) [reply src/components/CarouselItem.js:36]: `observeWidth` just observes the component for a short time period (from mounting until complete loading of the images). Then the observer is stopped (no battery drain on iPhone 5 as it doesn't support the ResizeObserver natively). When the view port is resized (e.g. turn the phone around) this method ensures, that the size is updated.

File: src/components/Carousel.js (hunk 1 above)
Comment (bunysae) [reply src/components/Carousel.js:493]: This method isn't needed in the last version. I filed PR #547 to kick it out.

**Comment by piotr-s-brainhub** [comment]
It's still working.

**Pull request merged**